contracts/command-batcher/src/encoding/evm.rs

```diff
@@ -0,0 +1,467 @@
+                msg.source_address,
+                ethereum_types::Address::from_str(&msg.destination_address).map_err(|e| {
+                    ContractError::InvalidMessage {
+                        reason: format!("destination_address is not a valid EVM address: {}", e),
+                    }
+                })?,
+                msg.payload_hash.as_slice().try_into().map_err(|e| {
+                    ContractError::InvalidMessage {
+                        reason: format!(
+                            "payload_hash length is not a valid keccak256 hash length: {}",
+                            e
+                        ),
+                    }
+                })?,
+            ),
+            id: command_id(msg.id),
+        })
+    }
+}
+
+impl CommandBatch {
+    pub fn new(
+        messages: Vec<Message>,
+        destination_chain_id: Uint256,
+    ) -> Result<Self, ContractError> {
+        let message_ids: Vec<String> = messages.iter().map(|msg| msg.id.clone()).collect();
+
+        let data = Data {
+            destination_chain_id,
+            commands: messages
+                .into_iter()
+                .map(|msg| msg.try_into())
+                .collect::<Result<Vec<Command>, ContractError>>()?,
+        };
+
+        let id = batch_id(&message_ids);
+
+        Ok(Self {
+            id,
+            message_ids,
+            data,
+        })
+    }
+
+    pub fn msg_to_sign(&self) -> HexBinary {
+        let msg = Keccak256::digest(self.data.encode().as_slice());
+
+        // Prefix for standard EVM signed data https://eips.ethereum.org/EIPS/eip-191
+        let unsigned = [
+            "\x19Ethereum Signed Message:\n32".as_bytes(), // Keccek256 hash length = 32
+            msg.as_slice(),
+        ]
+        .concat();
+
+        Keccak256::digest(unsigned).as_slice().into()
+    }
+
+    pub fn encode_execute_data(&self, multisig: Multisig) -> Result<HexBinary, ContractError> {
+        Ok(ethabi::encode(&[
+            Token::Bytes(self.data.encode().into()),
+            Token::Bytes(self.encode_proof(multisig)?.into()),
+        ])
+        .into())
+    }
+
+    fn encode_proof(&self, multisig: Multisig) -> Result<HexBinary, ContractError> {
+        let operators = sorted_operators(multisig.signers)?;
+
+        let (addresses, weights, signatures) = operators.iter().fold(
+            (vec![], vec![], vec![]),
+            |(mut addresses, mut weights, mut signatures), operator| {
+                addresses.push(Token::Address(ethereum_types::Address::from_slice(
+                    operator.address.as_slice(),
+                )));
+                weights.push(Token::Uint(ethereum_types::U256::from_big_endian(
+                    &operator.weight.to_be_bytes(),
+                )));
+
+                if let Some(signature) = &operator.signature {
+                    signatures.push(Token::Bytes(signature.into()));
+                }
+
+                (addresses, weights, signatures)
```

Comment:
I don't think this is right based on the code at https://github.com/axelarnetwork/axelar-cgp-solidity/blob/9c7a260c848011f27d6e7ecb1cba88de79206ccc/contracts/auth/AxelarAuthWeighted.sol#L90. Should only add address and weight to their corresponding Vec if signature is present.

Reply:
actually we need all of the addresses, but only a subset of signatures

Reply:
you are right